### ISSUE TITLE
ClusteringRaft.Fuzzer unit test failing

### ISSUE THREAD

**Issue by Tryneus** [rethinkdb/rethinkdb#4457]
The raft fuzzer test sometimes fails when writing to the cluster following the fuzzing:

```cpp
Version: rethinkdb 2.0.0-1-1239-g82f636-dirty (debug) (GCC 4.6.3)
error: Error in ../src/unittest/clustering_raft.cc at line 736:
error: Guarantee failed: [res]
error: Backtrace:
error: Wed Jun 24 16:20:08 2015

       1: rethinkdb_backtrace(void**, int) at rethinkdb_backtrace.cc:101
       2: backtrace_t::backtrace_t() at backtrace.cc:203
       3: lazy_backtrace_formatter_t::lazy_backtrace_formatter_t() at backtrace.cc:283
       4: format_backtrace(bool) at backtrace.cc:198
       5: report_fatal_error(char const*, int, char const*, ...) at errors.cc:83
       6: unittest::raft_fuzzer_t::run(unsigned long) at clustering_raft.cc:736
       7: unittest::run_fuzzer(long) at clustering_raft.cc:884
       8: void std::_Bind<void (*(std::_Placeholder<1>))(long)>::__call_c<void, long&, 0>(std::tuple<long&>&&, std::_Index_tuple<0>) const at functional:1156
       9: void std::_Bind<void (*(std::_Placeholder<1>))(long)>::operator()<long&, void>(long&) const at functional:1220
       10: pmap_runner_one_arg_t<std::_Bind<void (*(std::_Placeholder<1>))(long)>, long>::operator()() at pmap.hpp:21
       11: void pmap<std::_Bind<void (*(std::_Placeholder<1>))(long)> >(long, long, std::_Bind<void (*(std::_Placeholder<1>))(long)> const&) at pmap.hpp:48
       12: void pmap<std::_Bind<void (*(std::_Placeholder<1>))(long)> >(long, std::_Bind<void (*(std::_Placeholder<1>))(long)> const&) at pmap.hpp:54
       13: unittest::run_ClusteringRaft_Fuzzer() at clustering_raft.cc:889
       14: std::_Function_handler<void (), void (*)()>::_M_invoke(std::_Any_data const&) at functional:1780
       15: std::function<void ()>::operator()() const at functional:2162
       16: starter_t::run_wrapper(std::function<void ()> const&) at runtime.cc:61
       17: std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)>::operator()(starter_t*, std::function<void ()> const&) const at functional:551
       18: void std::_Bind<std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)> (starter_t*, std::function<void ()>)>::__call<void, , 0, 1>(std::tuple<>&&, std::_Index_tuple<0, 1>) at functional:1147
       19: void std::_Bind<std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)> (starter_t*, std::function<void ()>)>::operator()<, void>() at functional:1206
       20: std::_Function_handler<void (), std::_Bind<std::_Mem_fn<void (starter_t::*)(std::function<void ()> const&)> (starter_t*, std::function<void ()>)> >::_M_invoke(std::_Any_data const&) at functional:1780
       21: std::function<void ()>::operator()() const at functional:2162
       22: callable_action_instance_t<std::function<void ()>&>::run_action() at callable_action.hpp:31
       23: callable_action_wrapper_t::run() at runtime_utils.cc:43
       24: coro_t::run() at coroutines.cc:207
       25: coro_t* coro_t::spawn_sometime<std::function<void ()>&>(std::function<void ()>&) at coroutines.hpp:59
       26: starter_t::on_thread_switch() at runtime.cc:57
       27: linux_message_hub_t::on_event(int) at message_hub.cc:154
       28: epoll_event_queue_t::run() at epoll.cc:115
       29: linux_thread_pool_t::start_thread(void*) at thread_pool.cc:166
       30: /lib/x86_64-linux-gnu/libpthread.so.0(+0x7e9a) [0x7fb2386f7e9a] at 0x7fb2386f7e9a (/lib/x86_64-linux-gnu/libpthread.so.0)
       31: clone+0x6d at 0x7fb2384248bd (/lib/x86_64-linux-gnu/libc.so.6)
error: Exiting.
```

**Comment by timmaxw** [comment]
A partial solution is in CR 3020 by @Tryneus. There's still another bug, though, which I'll fix in a separate code review.

**Comment by timmaxw** [comment]
Further fixes are in CR 3022 by @VeXocide. After those fixes are applied, the fuzzer seems to pass consistently.

**Comment by timmaxw** [comment]
Further fixes merged in af5dba4db7d408d1d730ba003051fecbd28605b0.
